controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -350,64 +353,109 @@ public void refresh() {
-                                                    final long timestamp,
-                                                    final long leaseExpiryTime,
-                                                    final long maxExecutionExpiryTime,
-                                                    final long scaleGracePeriod) {
-        CompletableFuture<Integer> future = new CompletableFuture<>();
-        createNewTransactionNode(txId, timestamp, leaseExpiryTime, maxExecutionExpiryTime, scaleGracePeriod)
-                .whenComplete((value, ex) -> {
-                    if (ex != null) {
-                        if (Exceptions.unwrap(ex) instanceof StoreException.DataNotFoundException) {
-                            Futures.completeAfter(() -> createNewTransactionNode(txId, timestamp, leaseExpiryTime,
-                                    maxExecutionExpiryTime, scaleGracePeriod), future);
-                        } else {
-                            future.completeExceptionally(ex);
-                        }
+    CompletableFuture<Void> createNewTransaction(final UUID txId,
+                                                 final long timestamp,
+                                                 final long leaseExpiryTime,
+                                                 final long maxExecutionExpiryTime,
+                                                 final long scaleGracePeriod) {
+        long epoch = txId.getMostSignificantBits();
+        final String activePath = getActiveTxPath(epoch, txId.toString());
+        final byte[] txnRecord = new ActiveTxnRecord(timestamp, leaseExpiryTime, maxExecutionExpiryTime,
+                scaleGracePeriod, TxnStatus.OPEN).toByteArray();
+        // Note: this can throw DataNotFoundException as the epoch node (parent) may have been deleted.
+        return store.createZNodeIfNotExist(activePath, txnRecord, false)
+                .thenApply(x -> cache.invalidateCache(activePath));
+    }
+
+    // region epoch unique id generator
+    /**
+     * An epoch unique id generator generates a new long value which is unique for the epoch.
+     * This generator has requirement to be efficient and correct. To ensure correctness we will use a monotonically
+     * increasing counter which will return a new long. This will ensurue that we will never have collisions and values
+     * are unique.
+     * Several threads/processes could simultaneously try to generate new ids so the scheme should ensure that there are
+     * minimal concurrency penalties.
+     * Hence using curator's distributed atomic long recipe Or using a znoded and storing a long inside it and incrementing
+     * with CAS would be inefficient as there could be lots of write conflicts.
+     * So we have broken down the long counter into two integer counters.
+     * We use an least significant bigs(LSB) counter that performs count by using the znode.stat.version. Everytime the counter has to be incremented, we will
```

Comment:
Any znode update in zookeeper is equally expensive. The additional conditional check does not contribute any important additional overhead.

It is also not clear why we can't allocate a block of lsb values instead of doing a zk update on every new transaction id. If we do it this way, then transaction creation is limited by the write throughput of zk. It would be better to not have such a limitation.

Reply:
> Any znode update in zookeeper is equally expensive. The additional conditional check does not contribute any important additional overhead.

The overhead is borne by application which has to deal with write conflict and retry. So with more threads trying to concurrently update the same node, the overall time for each of those updates to go through will be pretty high. 

> It is also not clear why we can't allocate a block of lsb values instead of doing a zk update on every new transaction id. 

Blocks of lsb per epoch per stream distributed across controller instances will be very costly. We would ideally not want to maintain such distribution per stream in each controller instances memory for all streams.
We can have common blocks for all streams and all epochs though. For example say each controller instance gets a separate block range of 64bits long keyspace. So lets say we divide the entire space into predetermined 100 blocks. these blocks are distributed among all available controller instances.  
For all transactions across all streams, each controller receiving the request can choose the next id from the block. So this will make transaction ids across streams unique too (though that is not really a requirement). 

It will be simpler to implement too. But everytime a new assignment happens, we will need to persist it somewhere such that upon controller failure we dont reuse it. 
So a zookeeper update per assignment is inevitable, but it will make the implementation more efficient and simpler.